mlir/include/mlir/IR/OpDefinition.h

```diff
@@ -1053,6 +938,109 @@ struct hasSingleBlockImplicitTerminator<Op, false> {
+    assert(!region.empty() && "unexpected empty region");
+    return &region.front();
+  }
+  Region &getBodyRegion(unsigned idx = 0) {
+    return this->getOperation()->getRegion(idx);
+  }
+
+  //===------------------------------------------------------------------===//
+  // Single Region Utilities
+  //===------------------------------------------------------------------===//
+
+  /// The following are a set of methods only enabled when the parent
+  /// operation has a single region. Each of these methods take an additional
+  /// template parameter that represents the concrete operation so that we
+  /// can use SFINAE to disable the methods for non-single region operations.
+  template <typename OpT, typename T = void>
+  using enable_if_single_region =
+      std::enable_if_t<OpT::template hasTrait<OneRegion>(), T>;
+
+  template <typename OpT = ConcreteType>
+  enable_if_single_region<OpT, Block::iterator> begin() {
+    return getBody()->begin();
+  }
+  template <typename OpT = ConcreteType>
+  enable_if_single_region<OpT, Block::iterator> end() {
+    return getBody()->end();
+  }
+  template <typename OpT = ConcreteType>
+  enable_if_single_region<OpT, Operation &> front() {
+    return *begin();
+  }
+
+  /// Insert the operation into the back of the body right before the implicit
+  /// terminator.
+  template <typename OpT = ConcreteType,
+            std::enable_if_t<hasSingleBlockImplicitTerminator<OpT>::value,
+                             bool> = true>
+  enable_if_single_region<OpT> push_back(Operation *op) {
+    insert(Block::iterator(getBody()->getTerminator()), op);
+  }
+  /// Insert the operation into the back of the body.
+  template <typename OpT = ConcreteType,
+            std::enable_if_t<!hasSingleBlockImplicitTerminator<OpT>::value,
+                             bool> = true>
+  enable_if_single_region<OpT> push_back(Operation *op) {
+    insert(Block::iterator(getBody()->end()), op);
+  }
+
+  /// Insert the operation at the given insertion point.
+  template <typename OpT = ConcreteType>
+  enable_if_single_region<OpT> insert(Operation *insertPt, Operation *op) {
+    insert(Block::iterator(insertPt), op);
+  }
+  template <typename OpT = ConcreteType,
+            std::enable_if_t<hasSingleBlockImplicitTerminator<OpT>::value,
+                             bool> = true>
+  enable_if_single_region<OpT> insert(Block::iterator insertPt, Operation *op) {
+    // If expecting an implicit terminator, insert the operation before it.
+    Block *body = getBody();
+    if (insertPt == body->end())
+      insertPt = Block::iterator(body->getTerminator());
+    body->getOperations().insert(insertPt, op);
+  }
+  template <typename OpT = ConcreteType,
+            std::enable_if_t<!hasSingleBlockImplicitTerminator<OpT>::value,
+                             bool> = true>
+  enable_if_single_region<OpT> insert(Block::iterator insertPt, Operation *op) {
+    getBody()->getOperations().insert(insertPt, op);
+  }
```

Comment:
Specializing with `std::enable_if_t<!hasSingleBlockImplicitTerminator<OpT>::value, bool> = true` is my contribution here